doc/tutorials/rtvm_performance_tips.rst

```diff
@@ -0,0 +1,213 @@
+
+ACRN uses VMX_TSC_OFFSET_FULL as the offset between vTSC_ADJUST and
+pTSC_ADJUST; therefore, if VMX_TSC_OFFSET_FULL is zero, intercepting
+TSC_ADJUST and TSC_DEADLINE is not necessary. Otherwise, they should be
+intercepted to guarantee functionality.
+
+**Tip 6:** Utilize Preempt-RT Linux mechanisms to reduce the access of ICR
+from the RT core:
+
+#. Add "domain" to the "isolcpus" ( “isolcpus=nohz,domain,1” ) to the kernel parameters.
+#. Add "idle=poll" to the kernel parameters.
+#. Add "rcu_nocb_poll" along with "rcu_nocbs=1" to the kernel parameters.
+#. Disable the logging service like journald, syslogd if possible.
+
+These parameters are recommended for the guest Preempt-RT Linux. For a UP
+RTVM, ICR interception is not a problem. But for an SMP RTVM, IPI may be
+needed between vCPUs; these tips are about to reduce the ICR access. The
+example above assumes it is a dual-core RTVM, while core 0 is a housekeeping
+core and core 1 is a real-time core. The "domain" flag makes strong
+isolation of the RT core from the general SMP balancing and scheduling
+algorithms. "idle=poll" and "rcu_nocb_poll" could prevent the RT core from
+sending reschedule IPI to wakeup tasks on core 0 in most cases. And the
+disabling of the logging service is because an IPI may be issued to the
+housekeeping core to notify the logging service when there are kernel
+messages output on the RT core.
+
+.. note::
+   If an ICR access is inevitable within the RT critical section, please be
+   aware of the extra 3~4 us latency from each access.
+
+**TIP 7:** Create and initialize the RT tasks at the beginning to avoid
+runtime access to control registers.
+
+The access to Control Registers is another cause of a VM-exit. An ACRN access
+to CR3 and CR8 do not cause a VM-exit, but writes to CR0 and CR4 may cause a
+VM-exit, which would happen at the spawning and initialization of a new task.
+
+Isolating the impact of neighbor VMs
+************************************
+
+ACRN makes use of several technologies and hardware features to avoid the
+impact to the RTVM from neighbor VMs:
+
+**TIP 8:** Do not share CPUs allocated to the RTVM with other RT/non-RT VMs.
+
+ACRN enables CPU sharing to improve the utilization of CPU resources.
+However, for RT VM, CPUs should be dedicatedly allocated for the determinism.
+
+**TIP 9:** Use RDT such as CAT and MBA to allocate dedicated resources to
+the RTVM.
+
+ACRN enables the Intel® Resource Director Technology, such as CAT and MBA,
+components such as the GPU via memory hierarchy. The availability of RDT is
+hardware-specific. Refer to the :ref:`rdt_configuration`.
+
+**TIP 10:** Lock the GPU to a feasible lowest frequency.
+
+GPU can put heavy pressure on the power/memory subsystem, so locking the GPU
+frequency as low as possible can help to improve the determinism of RT
+performance. It can be locked in the BIOS, but the availability of certain
+BIOS option is platform-specific.

```

Comment:
I believe there are user-space tools that can be used in the Service VM to do this as well, should the Bios not support it (`intel_gpu_frequency`, and possibly by directly tweaking some files in `/sys` too?). Can we add this?